README.md
# matrixTest

Tests for block-diagonal structure in symmetric matrices (e.g. correlation matrices) under the null hypothesis of exchangeable off-diagonal elements. As described in Segal et al. (2019), these tests can be useful for construct validation either by themselves or as a complement to confirmatory factor analysis. Monte Carlo methods are used to approximate the permutation p-value with Hubert's Gamma (Hubert, 1976) and a t-statistic with unequal variance. This package also implements the Chi-squared statistic described by Steiger (1980).

## Installation

```{r}
library("devtools")
install_github("bdsegal/matrixTest")
```

## Examples

```{r}
library(matrixTest)
library(ggplot2)
library(reshape2)
     
# prepare data for matrixTest -------------------------------------------------
data("big5")

# get column numbers for questionnaire items
items <- grep("[0-9]", colnames(big5))

# compute Spearman's correlation matrix
A <- cor(big5[, items], use = "complete.obs", method = "spearman")

# Specify groups
groups <- "extrovert ~ E1 + E2 + E3 + E4 + E5 + E6 + E7 + E8 + E9 + E10
           neurotic ~ N1 + N2 + N3 + N4 + N5 + N6 + N7 + N8 + N9 + N10
           agreeable ~ A1 + A2 + A3 + A4 + A5 + A6 + A7 + A8 + A9 + A10
           conscientious ~ C1 + C2 + C3 + C4 + C5 + C6 + C7 + C8 + C9 + C10
           open ~ O1 + O2 + O3 + O4 + O5 + O6 + O7 + O8 + O9 + O10"
     
# compute permutation p-values ------------------------------------------------
result <- matrixTest(A = A, groups = groups, B = 1000, absolute = TRUE)
result

# Visualize groups ------------------------------------------------------------
ord <- unlist(result$group_list)
diag(A) <- NA # remove diagonals from color scale
Am <- melt(A[ord, ord])
names(Am) <- c("x", "y", "value")
Am$y <- factor(Am$y, levels = rev(levels(Am$y)))

ggplot(aes(x = x, y = y, fill = abs(value)), data = Am)+
  geom_tile()+
  theme_bw(18)+
  scale_fill_gradient2(space="Lab", name="abs(Cor)", lim = c(0, 1))+
  labs(x = "", y = "")+
  theme(axis.text.x = element_text(angle = 90, vjust = .35,hjust=1))
```

## References

Hubert, L., Schultz, J. (1976) Quadratic assignment as a general data analysis strategy. British Journal of Mathematical and Statistical Psychology, 29(2):190–241.

Steiger, J. H. (1980). Tests for comparing elements of a correlation matrix. Psychological Bulletin, 87(2):245–251.

Segal, B. D., Braun, T., Gonzalez, R., and Elliott, M. R. (2019). Tests of matrix structure for construct validation. Psychometrika, 84(1), 65-83.


NEWS.md
# matrixText 1.0.0

This is the initial release of matrixTest.

R/data.R
#' Big Five personality questionnaire
#'
#' A dataset containing answers to a Big Five Personality Test conducted by
#' http://personality-testing.info. These data were collected (c. 2012) through
#' an interactive online personality test. The test was constructed with items 
#' from the International Personality Item Pool. Participants were informed that
#' their responses would be recorded and used for research at the beginning of 
#' the test and asked to confirm their consent at the end of the test. The items
#' were rated on a five point scale where 1=Disagree, 3=Neutral, 5=Agree. 
#' All were presented on one page in the order E1, N1, A1, C1, O1, E2,...
#'
#' This dataset is for demonstration purposes only. Please see
#' http://personality-testing.info/privacypolicy.html and 
#' http://personality-testing.info/about for more information.
#'
#' @format A data frame with 19,719 rows of  57 variables:
#' \describe{
#'   \item{race}{1=Mixed Race, 2=Arctic (Siberian, Eskimo), 
#'               3=Caucasian (European), 4=Caucasian (Indian), 
#'               5=Caucasian (Middle East), 6=Caucasian (North African, Other), 
#'               7=Indigenous Australian, 8=Native American, 
#'               9=North East Asian (Mongol, Tibetan, Korean Japanese, etc), 
#'               10=Pacific (Polynesian, Micronesian, etc), 
#'               11=South East Asian (Chinese, Thai, Malay, Filipino, etc), 
#'               12=West African, Bushmen, Ethiopian, 13=Other}
#'  \item{age}{Entered as text (individuals reporting age < 13 were not recorded)}
#'  \item{engnat}{Response to "is English your native language?". 1=yes, 2=no}
#'  \item{gender}{1=Male, 2=Female, 3=Other}
#'  \item{hand}{"What hand do you use to write with?". 1=Right, 2=Left, 3=Both}
#'  \item{country}{The participant's technical location. ISO country code}
#'  \item{source}{How the participant came to the test. Based on HTTP Referer. 
#'                1=from another page on the test website, 2=from google, 
#'                3=from facebook, 4=from any url with ".edu" in its domain name,
#'                6=other source, or HTTP Referer not provided}
#'  \item{E1}{I am the life of the party.}
#'  \item{E2}{I don't talk a lot.}
#'  \item{E3}{I feel comfortable around people.}
#'  \item{E4}{I keep in the background.}
#'  \item{E5}{I start conversations.}
#'  \item{E6}{I have little to say.}
#'  \item{E7}{I talk to a lot of different people at parties.}
#'  \item{E8}{I don't like to draw attention to myself.}
#'  \item{E9}{I don't mind being the center of attention.}
#'  \item{E10}{I am quiet around strangers.}
#'  \item{N1}{I get stressed out easily.}
#'  \item{N2}{I am relaxed most of the time.}
#'  \item{N3}{I worry about things.}
#'  \item{N4}{I seldom feel blue.}
#'  \item{N5}{I am easily disturbed.}
#'  \item{N6}{I get upset easily.}
#'  \item{N7}{I change my mood a lot.}
#'  \item{N8}{I have frequent mood swings.}
#'  \item{N9}{I get irritated easily.}
#'  \item{N10}{I often feel blue.}
#'  \item{A1}{I feel little concern for others.}
#'  \item{A2}{I am interested in people.}
#'  \item{A3}{I insult people.}
#'  \item{A4}{I sympathize with others' feelings.}
#'  \item{A5}{I am not interested in other people's problems.}
#'  \item{A6}{I have a soft heart.}
#'  \item{A7}{I am not really interested in others.}
#'  \item{A8}{I take time out for others.}
#'  \item{A9}{I feel others' emotions.}
#'  \item{A10}{I make people feel at ease.}
#'  \item{C1}{I am always prepared.}
#'  \item{C2}{I leave my belongings around.}
#'  \item{C3}{I pay attention to details.}
#'  \item{C4}{I make a mess of things.}
#'  \item{C5}{I get chores done right away.}
#'  \item{C6}{I often forget to put things back in their proper place.}
#'  \item{C7}{I like order.}
#'  \item{C8}{I shirk my duties.}
#'  \item{C9}{I follow a schedule.}
#'  \item{C10}{I am exacting in my work.}
#'  \item{O1}{I have a rich vocabulary.}
#'  \item{O2}{I have difficulty understanding abstract ideas.}
#'  \item{O3}{I have a vivid imagination.}
#'  \item{O4}{I am not interested in abstract ideas.}
#'  \item{O5}{I have excellent ideas.}
#'  \item{O6}{I do not have a good imagination.}
#'  \item{O7}{I am quick to understand things.}
#'  \item{O8}{I use difficult words.}
#'  \item{O9}{I spend time reflecting on things.}
#'  \item{O10}{I am full of ideas.}
#' }
#' @source \url{http://personality-testing.info/_rawdata/}
"big5"

R/pearson_filon.R
# Testing using Pearson and Filon's results -- Steiger's method

sigmaRhoFun <- function(j, k, h, m, A) {
  #' Utility function   
  #'
  #' @param j First index
  #' @param k Second index
  #' @param h Third index
  #' @param m Fourth index
  #' @param A Correlation matrix
  #' @export

  0.5 * (
  (A[j,h] - A[j,k]*A[k,h]) * (A[k,m] - A[k,h]*A[h,m]) +
  (A[j,m] - A[j,h]*A[h,m]) * (A[k,h] - A[k,j]*A[j,h]) +
  (A[j,h] - A[j,m]*A[m,h]) * (A[k,m] - A[k,j]*A[j,m]) +
  (A[j,m] - A[j,k]*A[k,m]) * (A[k,h] - A[k,m]*A[m,h]))
}

sigmaZFun <- function(s, t, index, A, Sigma) {
  #' Utility function
  #'
  #' @param s First index
  #' @param t Second index
  #' @param index Matrix with two columns with index pairs given by rows
  #' @param A Correlation matrix
  #' @param Sigma Variance-covariances of correlation matrix A
  #' @export
  Sigma[s, t] / ((1 - A[index[s, 1], index[s, 2]]^2) * 
                 (1 - A[index[t, 1], index[t, 2]]^2))
}


X2Fun <- function(data, group_list, corMethod = "spearman"){
  #' Goodness-of-fit chi-squared statistic described by Steiger (1980).
  #'
  #' References:
  #' Steiger, J. H. (1980). Tests for comparing elements of a correlation matrix. Psychological Bulletin, 87(2) 245-251.
  #' @param data Data frame
  #' @param group_list List of column indices of A for each group
  #' @param corMethod Type of correlations; passed to cor()
  #' @importFrom stats complete.cases cor pchisq t.test
  #' @export

  data <- data[complete.cases(data), ]
  N <- nrow(data)
  
  A <- cor(data, method = corMethod)

  p <- ncol(A)
  K <- length(group_list)
  
  if(sum(sapply(group_list,length)) > p){
    warning("Some variables may be assigned to multiple clusters")
  } else if(sum(sapply(group_list,length)) < p){
    warning("Some variables not assigned to a cluster")
  }
  
  # create delta matrix
  Delta <- matrix(nrow=p,ncol=p)
  for (i in 1:p){
    for (j in 1:p){
      Delta[i,j] <- deltaSub(i, j, group_list)
    }
  }
  
  # vectors of first and second indices in correlation matrix
  first <- NULL
  second <- NULL
  for (j in 2:p) {
  for (i in 1:(j-1)) {
    first <- c(first, i)
    second <- c(second, j)
    }
  }
  
  index <- cbind(first, second)
  nInd <- nrow(index)

  L <- cbind(1, Delta[upper.tri(Delta)])
  rho <- A[upper.tri(A)]

  # get initial OLS estimate and put in matrix for easy processing
  rhoHatOLSTri <- matrix(NA, nrow = nrow(A), ncol = ncol(A))
  rhoHatOLSTri[upper.tri(rhoHatOLSTri)] <- L %*% solve(crossprod(L, L), crossprod(L, rho))
  rhoHatOLSTri[lower.tri(rhoHatOLSTri)] <- t(rhoHatOLSTri)[lower.tri(rhoHatOLSTri)]
  diag(rhoHatOLSTri) <- 1

  # get SigmaOLS
  SigmaOLS <- matrix(0, nrow = nInd, ncol = nInd)
  for(s in 1:(nInd -1)) {
    for (t in (s):nInd) {
      SigmaOLS[s,t] <- sigmaRhoFun(index[s,1], index[s,2], 
                                   index[t,1], index[t,2], rhoHatOLSTri)
    }
  }
  SigmaOLS[lower.tri(SigmaOLS)] <- t(SigmaOLS)[lower.tri(SigmaOLS)]

  rhoHatGLS <- L %*% solve(crossprod(L, solve(SigmaOLS, L)), crossprod(L, 
               solve(SigmaOLS, rho)))

  # get covariance for Z transformed variables
  SigmaZ <- matrix(0, nrow = nInd, ncol = nInd)
  for (s in 1:(nInd - 1)) {
    for (t in s:nInd) {
      SigmaZ[s, t] <- sigmaZFun(s, t, index, A, SigmaOLS)
    }
  }
  SigmaZ[lower.tri(SigmaZ)] <- t(SigmaZ)[lower.tri(SigmaZ)]

  X2 <- (N - 3) *t(atanh(rho) -atanh(rhoHatGLS)) %*% 
        solve(SigmaZ, atanh(rho) - atanh(rhoHatGLS))
  df <- p*(p-1)/2 - 2

  result <- c(X2, df, pchisq(X2, df = df, lower.tail = FALSE))
  names(result) <- c("X2", "df", "pval")

  return(result)
}


R/hubert_t.R
# MC estimates of permutation p-values with Hubert's Gamma and
# t-statistic with unequal variance

# functions for making Delta matrix -------------------------------------------

deltaSub <- function(i,j, group_list){
  #' Utility function
  #'
  #' This function outputs 1 if i and j are in at least one group together, 
  #' and 0 otherwise
  #' @param i First index
  #' @param j Second index
  #' @param group_list List of indices for each block
  #' @export

  i_group <- which(sapply(group_list, function(group){i %in% group}))
  j_group <- which(sapply(group_list, function(group){j %in% group}))
  
  # allows for multiple membership
  as.numeric(length(intersect(i_group, j_group)) >= 1)
}

multiSub <- function(i, j, group){
  #' Utility function
  #'
  #' This function outputs TRUE if either i or j are in group, FALSE otherwise
  #' @param i First index
  #' @param j Second index
  #' @param group Indices for items in group
  #' @export

  i_in <- i %in% group
  j_in <- j %in% group
  
  (i_in + j_in) >=1
}

# Both t-stat and Hubert's gamma ----------------------------------------------
matrixTestSub <- function(A, group_list_ord, Delta, multi_group_ind,
                         A_upper_ind, K){
  #' Utility function: Gamma and t
  #' @param A Distance or similarity matrix, e.g. correlation
  #' @param group_list_ord List of groupings for ordered matrix A
  #' @param Delta Delta matrix
  #' @param multi_group_ind List of indicator matrices for membership in block k test
  #' @param A_upper_ind indicator matrix for upper triangular elements
  #' @param K Total number of hypothesized blocks
  #' @return Gamma_overall: Overall Hubert's Gamma
  #' @return Gamma_multi: Block-specific Hubert's Gamma
  #' @return t_overall: Overall t-statistic (unequal variance)
  #' @return t_multi: Block-specific t-statistics
  #' @return Ak_list: List of values in A used for each block-specific test
  #' @return Deltak_list: List of values in Delta used for each block-specific test
  #' @return A_upper: Upper triangular elements of A (used for box plot function)
  #' @return Delta_upper: Upper triangular elements of Delta (used for box plot function)
  #' @export

  # Calculate Hubert's Gamma with upper triangular elements
  A_upper <- A[upper.tri(A)]
  Delta_upper <- Delta[upper.tri(Delta)]

  Gamma_overall <- cor(A_upper, Delta_upper)

  g1 <- A_upper[which(Delta_upper == 1)]
  g0 <- A_upper[which(Delta_upper == 0)]
  t_overall <- t.test(g1, g0, equal.var = FALSE)$statistic

  # multiple hypothesis testing ---------------------------
  Ak_list <- list()
  Deltak_list <- list()

  Gamma_multi <- rep(NA, K)
  t_multi <- rep(NA, K)
  for (k in 1:K){

   Ak_list[[k]] <- A[A_upper_ind & multi_group_ind[[k]]]
   Deltak_list[[k]] <- Delta[A_upper_ind & multi_group_ind[[k]]]
   Gamma_multi[k] <- cor(Ak_list[[k]], Deltak_list[[k]])

   g0 <- Ak_list[[k]][which(Deltak_list[[k]] == 0)]
   g1 <- Ak_list[[k]][which(Deltak_list[[k]] == 1)]
   t_multi[k] <- t.test(g1, g0, equal.var = FALSE)$statistic

  }

  return(list(Gamma_overall = Gamma_overall,
              Gamma_multi = Gamma_multi,
  	          t_overall = t_overall,
              t_multi = t_multi,
    				  Ak_list = Ak_list,
              Deltak_list = Deltak_list,
    				  A_upper = A_upper,
              Delta_upper = Delta_upper))
}

makeGroupList <- function(groups, A) {
  #' Utility function: Turn character string in lavaan syntax into list
  #'
  #' @param groups Character string in lavaan syntax specifying groups
  #' @param A A Distance or similarity matrix. Must have column names
  #'
  #' @return group_list List of column indices of A for each group
  #' @export

  split <- gsub(" ", "", unlist(strsplit(groups, "\\n")))
  group_list <- vector(length = length(split), mode = "list")
  for (i in 1:length(split)) {
    sub_split <- unlist(strsplit(split[i], "~"))
    names(group_list)[i] <- sub_split[1]

    labels <- unlist(strsplit(sub_split[2], "\\+"))
    group_list[[i]] <- which(colnames(A) %in% labels)
  }

  rm_group <- which(is.na(names(group_list)))
  if (length(rm_group) >= 1) {
    group_list <- group_list[-rm_group]
  }

return(group_list)
}

matrixTest <- function(A, group_list = NULL, groups = NULL, B = 1000, absolute = TRUE){
  #' Gamma and t function
  #'
  #' This function computes p-values for Hubert's Gamma and t statistics
  #' @param A Distance or similarity matrix, e.g. correlation
  #' @param group_list List of column indices of A for each group
  #' @param groups CFA model in lavaan syntax. 
  #' @param B Number of Monte Carlo resamples (defaults to B=1000)
  #' @param absolute Use the absolute values of A (defaults to TRUE)

  #' @return pt_overall_one_sided: Overall one-sided p-value using t statistic
  #' @return pt_overall_two_sided: Overall two-sided p-value using t statistic
  #' @return pt_multi_one_sided: Block-specific one-sided p-values using t statistic
  #' @return pt_multi_two_sided: Block-specific two-sided p-values using t statistic
  #' @return t0 Observed overall: t statistic
  #' @return t0k: Observed block-specific t statistic
  #' @return t_overall: Vector of overall t statistics from permuted A
  #' @return t_max_one_sided: Vector of max t statistics from permuted A (one-sided)
  #' @return t_max_two_sided: Vector of max t statistics from permuted A (two-sided)
  #' @return pG_overall_one_sided: Overall one-sided p-value using Hubert's Gamma
  #' @return pG_overall_two_sided: Overall two-sided p-value using Hubert's Gamma
  #' @return pG_multi_one_sided: Block-specific one-sided p-values using Hubert's Gamma
  #' @return pG_multi_two_sided: Block-specific two-sided p-values using Hubert's Gamma
  #' @return Gamma0: Observed overall Hubert's Gamma
  #' @return Gamma0k: Observed block-specific Hubert's Gamma
  #' @return Gamma_overall: Vector of Hubert's Gamma statistics from permuted A
  #' @return Gamma_max_one_sided: Vector of max Hubert's Gamma statistics from permuted A (one-sided)
  #' @return Gamma_max_two_sided: Vector of max Hubert's Gamma statistics from permuted A (two-sided)
  #' @return B: number of Monte Carlo resamples
  #' @return group_list: List of column/row indices corresponding to each group
  #' @keywords Hubert's Gamma, t-statistic, permutation, MC resampling
  #' @export
  #' @examples
  #' # example for matrixText package
  #' library(matrixTest)
  #' data("big5")
  #'
  #' # get column numbers for questionnaire items
  #' items <- grep("[0-9]", colnames(big5))
  #' 
  #' # compute Spearman's correlation matrix
  #' A <- cor(big5[, items], use = "complete.obs", method = "spearman")
  #' 
  #' # specify the groups
  #' groups <- "extrovert ~ E1 + E2 + E3 + E4 + E5 + E6 + E7 + E8 + E9 + E10
  #'            neurotic ~ N1 + N2 + N3 + N4 + N5 + N6 + N7 + N8 + N9 + N10
  #'            agreeable ~ A1 + A2 + A3 + A4 + A5 + A6 + A7 + A8 + A9 + A10
  #'            conscientious ~ C1 + C2 + C3 + C4 + C5 + C6 + C7 + C8 + C9 + C10
  #'            open ~ O1 + O2 + O3 + O4 + O5 + O6 + O7 + O8 + O9 + O10"
  #'
  #' # compute permutation p-values
  #' # Note: Using small B for fast execution. Set B >= 1000 in practice.
  #' result <- matrixTest(A = A, groups = groups, B = 100, absolute = TRUE)
  #'
  #' # Note: two-sided p-values from Hubert's Gamma printed by default
  #' #       other results available by directing accessing them from the
  #' #       returned object
  #' result
  #'
  #' # Alternative approach for specifying the groups as a list of column/row indices
  #' extrovert <- grep("E", colnames(A))
  #' neurotic <- grep("N", colnames(A))
  #' agreeable <- grep("A", colnames(A))
  #' conscientious <- grep("C", colnames(A))
  #' open <- grep("O", colnames(A))
  #' 
  #' # put blocks/groups in list
  #' group_list <- list(extrovert = extrovert, 
  #'                    neurotic = neurotic, 
  #'                    agreeable = agreeable, 
  #'                    conscientious = conscientious,
  #'                    open = open)
  #' 
  #' # Note: Using small B for fast execution. Set B >= 1000 in practice.
  #' result <- matrixTest(A = A, group_list = group_list, B = 100, absolute = TRUE)
  #'
  #' # Note: two-sided p-values from Hubert's Gamma printed by default
  #' #       other results available by directing accessing them from the
  #' #       returned object
  #' result
  #' 
  #' # Visualize groups
  #' library(ggplot2)
  #' library(reshape2)
  #' 
  #' ord <- unlist(result$group_list)
  #' diag(A) <- NA # remove diagonals from color scale
  #' Am <- melt(A[ord, ord])
  #' names(Am) <- c("x", "y", "value")
  #' Am$y <- factor(Am$y, levels = rev(levels(Am$y)))
  #' 
  #' ggplot(aes(x = x, y = y, fill = abs(value)), data = Am)+
  #'   geom_tile()+
  #'   theme_bw(18)+
  #'   scale_fill_gradient2(space="Lab", name="abs(Cor)", lim = c(0, 1))+
  #'   labs(x = "", y = "")+
  #'   theme(axis.text.x = element_text(angle = 90, vjust = .35,hjust=1))

  if(!is.null(groups) & !is.null(group_list)) {
    stop("Both a 'groups' and 'group_list' were supplied as arguments. Only one of the two can be given.")
  } else if(is.null(groups) & is.null(group_list)) {
    stop("Either 'groups' or 'group_list' must be supplied as an argument")
  }

  # if groups given, convert to group_list
  if(!is.null(groups)) {
    if(is.null(colnames(A))) {
      stop("When 'groups' is supplied, the matrix A must have column names")
    }

  group_list <- makeGroupList(groups, A)
  }

  if(absolute){
    A <- abs(A)
  }

  p <- ncol(A)
  K <- length(group_list)

  if(sum(sapply(group_list, length)) > p){
    warning("Some variables may be assigned to multiple clusters")
  } else if(sum(sapply(group_list, length)) < p){
    warning("Some variables not assigned to a cluster")
  }

  # clean up names of groups
  if(is.null(names(group_list))) {
    names(group_list) <- paste("block", 1:K)
  } else {
    no_name <- which(sapply(names(group_list), nchar) == 0)
    names(group_list)[no_name] <- paste("block", no_name)
  }

  # create delta matrix
  Delta <- matrix(nrow=p,ncol=p)
  for (i in 1:p){
    for (j in 1:p){
       Delta[i,j] <- deltaSub(i, j, group_list)
    }
  }

  # order matrices by hypothesized groups
  ord <- unlist(group_list)
  A_ord <- A[ord, ord]
  Delta_ord <- Delta[ord, ord]

  # get group list for reordered matrices
  nk <- sapply(group_list, length)
  firstIndex <- 1
  group_list_ord <- list()
  for (k in 1:K) {
    group_list_ord[[k]] <- firstIndex:(firstIndex+nk[k] -1)
    firstIndex <- firstIndex + nk[k]
  }

  # get indicator matrices for block-specific tests
  newOrd <- unlist(group_list_ord)
  A_upper_ind <- upper.tri(A)
  multi_group_ind <- list()
  for (k in 1:K){
    multi_group_ind[[k]] <- outer(newOrd, newOrd, multiSub, 
                                  group = group_list_ord[[k]])
  }

  # observed test statistic
  out0 <- matrixTestSub(A = A_ord,
  	                  group_list_ord = group_list_ord,
  	                  Delta = Delta_ord, 
   	                  multi_group_ind = multi_group_ind,
   	                  A_upper_ind = A_upper_ind,
   	                  K = K)

  Gamma0 <- out0$Gamma_overall
  Gamma0k <- out0$Gamma_multi
  t0 <- out0$t_overall
  t0k <- out0$t_multi

  # permutations
  Gamma_overall <- rep(NA, B)
  Gamma_k_mat <- matrix(nrow=B, ncol=K) 
  t_overall <- rep(NA, B)
  t_k_mat <- matrix(nrow=B, ncol=K)

  for (b in 1:B){
    perm <- sample(1:p)
    Ab <- A_ord[perm, perm]
    out <- matrixTestSub(A = Ab,
    	                 group_list_ord = group_list_ord,
    	                 Delta = Delta_ord,
     	                 multi_group_ind = multi_group_ind,
    	                 A_upper_ind = A_upper_ind,
     	                 K = K)
    Gamma_overall[b] <- out$Gamma_overall
    Gamma_k_mat[b,] <- out$Gamma_multi
    t_overall[b] <- out$t_overall
    t_k_mat[b,] <- out$t_multi
  }
  
  # t-statistic 
  pt_overall_one_sided <- mean(c(t0, t_overall) >= t0)
  pt_overall_two_sided <- mean(abs(c(t0, t_overall)) >= abs(t0))
  t_max_one_sided <- apply(t_k_mat, 1, max)
  t_max_two_sided <- apply(t_k_mat, 1, function(x){max(abs(x))})
  pt_multi_one_sided <- sapply(t0k, function(x){
                               mean(c(t_max_one_sided, x) >= x)})
  pt_multi_two_sided <- sapply(t0k, function(x){
                               mean(c(t_max_two_sided, abs(x)) >= abs(x))})
  names(pt_multi_one_sided) <- names(pt_multi_two_sided) <- names(group_list)
  names(t0k) <- colnames(t_k_mat) <- names(group_list)
   
  # Hubert's Gamma
  pG_overall_one_sided <- mean(c(Gamma0, Gamma_overall) >= Gamma0)
  pG_overall_two_sided <- mean(abs(c(Gamma0, Gamma_overall)) >= abs(Gamma0))
  Gamma_max_one_sided <- apply(Gamma_k_mat, 1, max)
  Gamma_max_two_sided <- apply(Gamma_k_mat, 1, function(x){max(abs(x))})
  pG_multi_one_sided <- sapply(Gamma0k, function(x){
                               mean(c(Gamma_max_one_sided, x) >= x)})
  pG_multi_two_sided <- sapply(Gamma0k, function(x){
                               mean(c(Gamma_max_two_sided, abs(x)) >= abs(x))})
  names(pG_multi_one_sided) <- names(group_list)
  names(Gamma0k) <- names(group_list)
   
  ret <- list(pt_overall_one_sided = pt_overall_one_sided,
              pt_overall_two_sided = pt_overall_two_sided,
              pt_multi_one_sided = pt_multi_one_sided,
              pt_multi_two_sided = pt_multi_two_sided,
              t0 = t0,
              t0k = t0k,
              t_overall = t_overall,
              t_max_one_sided = t_max_one_sided,
              t_max_two_sided = t_max_two_sided,
              pG_overall_one_sided = pG_overall_one_sided,
              pG_overall_two_sided = pG_overall_two_sided,
              pG_multi_one_sided = pG_multi_one_sided,
              pG_multi_two_sided = pG_multi_two_sided,
              Gamma0 = Gamma0,
              Gamma0k = Gamma0k,
              Gamma_overall = Gamma_overall,
              Gamma_max_one_sided = Gamma_max_one_sided,
              Gamma_max_two_sided = Gamma_max_two_sided,
              group_list = group_list,
              B = B)
  class(ret) <- "mt"
  return(ret)
}

print.mt <- function(x, ...) {
  #' Print function for results
  #'
  #' This function prints results from matrixTest
  #' @param x Output from matrixTest
  #' @param ... Further arguments passed to print
  #' @keywords matrixTest print
  #' @export

  if(x$pG_overall_two_sided > 1/(x$B + 1)) {
    result <- paste("    Test of matrix structure   \n\n",
      prettyNum(x$B, big.mark = ","), " Monte Carlo resamples, two-sided p-values",
      "\n\nOverall Hubert's Gamma",
      "\nGamma0 = ", signif(x$Gamma0, 3), ", p-val = ", signif(x$pG_overall_two_sided, 3),
      "\n\nBlock-specific Hubert's Gamma", sep = "")
  } else {
    result <- paste("    Test of matrix structure   \n\n",
      prettyNum(x$B, big.mark = ","), " Monte Carlo resamples, two-sided p-values",
      "\n\nOverall Hubert's Gamma",
      "\nGamma0 = ", signif(x$Gamma0, 3), ", p-val < ", 1/x$B,
      "\n\nBlock-specific Hubert's Gamma", sep = "")
  }

  p_multi <- x$pG_multi_two_sided
  hit_min <- p_multi == 1/(x$B + 1)
  p_multi <- as.character(signif(p_multi, 3))
  p_multi[hit_min] <- paste("<", 1/(x$B))
  writeLines(result)
  print(data.frame(Gamma0 = signif(x$Gamma0k, 3),
                   pval = p_multi),
        ...)
}


prepBoxPlots <- function(A, groups = NULL, group_list = NULL, absolute = TRUE){
  #' Prepare data for box plots
  #'
  #' This function prepares the data for making box plots
  #' @param A Distance or similarity matrix, e.g. correlation
  #' @param groups CFA model in lavaan syntax. 
  #' @param group_list List of groupings
  #' @param absolute Use the absolute values of A (defaults to TRUE)
  #' @return multi: data frame for making box plots for block-specific tests
  #' @return overall: data frame for making box plots for overall test
  #' @export
  #' @examples
  #' library(matrixTest)
  #' library(ggplot2)
  #' 
  #' data("big5")
  #' 
  #' # get column numbers for questionnaire items
  #' items <- grep("[0-9]", colnames(big5))
  #' 
  #' # compute Spearman's correlation matrix
  #' A <- cor(big5[, items], use = "complete.obs", method = "spearman")
  #' 
  #' groups <- "extrovert ~ E1 + E2 + E3 + E4 + E5 + E6 + E7 + E8 + E9 + E10
  #'            neurotic ~ N1 + N2 + N3 + N4 + N5 + N6 + N7 + N8 + N9 + N10
  #'            agreeable ~ A1 + A2 + A3 + A4 + A5 + A6 + A7 + A8 + A9 + A10
  #'            conscientious ~ C1 + C2 + C3 + C4 + C5 + C6 + C7 + C8 + C9 + C10
  #'            open ~ O1 + O2 + O3 + O4 + O5 + O6 + O7 + O8 + O9 + O10"
  #' 
  #' # Make box plots contrasting within and between group correlations
  #' box <- prepBoxPlots(A = A, groups = groups, absolute = TRUE)
  #' 
  #' ggplot(aes(x = as.factor(delta), y = a), data = box$overall)+
  #'   geom_boxplot()+
  #'   theme_bw(22)+
  #'   labs(x = expression(Delta), y="|a|")
  #' 
  #' dev.new(width = 12, height = 5)
  #' ggplot(aes(x = as.factor(delta), y = a), data = box$multi)+
  #'   geom_boxplot()+
  #'   facet_grid(~block)+
  #'   theme_bw(22)+
  #' labs(x = expression(Delta), y = "|a|")
   
  if(!is.null(groups) & !is.null(group_list)) {
    stop("Both a 'groups' and 'group_list' were supplied as arguments. Only one of the two can be given.")
  } else if(is.null(groups) & is.null(group_list)) {
    stop("Either 'groups' or 'group_list' must be supplied as an argument")
  }

  # if groups given, convert to group_list
  if(!is.null(groups)) {
    if(is.null(colnames(A))) {
      stop("When 'groups' is supplied, the matrix A must have column names")
    }

    group_list <- makeGroupList(groups, A)
  }

   if(absolute){
      A <- abs(A)
   }
   
   p <- ncol(A)
   K <- length(group_list)
   
   if(sum(sapply(group_list,length)) > p){
      warning("Some variables may be assigned to multiple clusters")
   } else if(sum(sapply(group_list,length)) < p){
      warning("Some variables not assigned to a cluster")
   }
   
  # clean up names of groups
  if(is.null(names(group_list))) {
    names(group_list) <- paste("block", 1:K)
  } else {
    no_name <- which(sapply(names(group_list), nchar) == 0)
    names(group_list)[no_name] <- paste("block", no_name)
  }

   # create delta matrix
   Delta <- matrix(nrow = p, ncol = p)
   for (i in 1:p){
      for (j in 1:p){
         Delta[i,j] <- deltaSub(i, j, group_list)
      }
   }

  # order matrices by hypothesized groups
  ord <- unlist(group_list)
  A_ord <- A[ord, ord]
  Delta_ord <- Delta[ord, ord]

  # get group list for reordered matrices
  nk <- sapply(group_list, length)
  firstIndex <- 1
  group_list_ord <- list()
  for (k in 1:K) {
    group_list_ord[[k]] <- firstIndex:(firstIndex + nk[k] -1)
    firstIndex <- firstIndex + nk[k]
  }

  # get indicator matrices for block-specific tests
  newOrd <- unlist(group_list_ord)
  A_upper_ind <- upper.tri(A)
  multi_group_ind <- list()
  for (k in 1:K){
    multi_group_ind[[k]] <- outer(newOrd, newOrd, multiSub, 
                                  group = group_list_ord[[k]])
  }
 
  # observed test statistic
  out <- matrixTestSub(A = A_ord,
                     group_list_ord = group_list_ord,
                     Delta = Delta_ord, 
                     multi_group_ind = multi_group_ind,
                     A_upper_ind = A_upper_ind,
                     K = K)
   
  box_overall <- data.frame(a = out$A_upper, delta = out$Delta_upper)
   
  box_list <- list()
  for (k in 1:K){
    box_list[[k]] <- data.frame(a = out$Ak_list[[k]], 
                                delta = out$Deltak_list[[k]],
                                block = names(group_list)[k])
  }
   
  box_multi <- do.call(rbind, box_list)
  
  return(list(multi = box_multi, overall = box_overall))
}
